File: leetcode/0020-valid-parentheses/solution.py
class Solution(object):
    def isValid(self, s):
        parenthesesStack = []
        parenthesesPairs = {")": "(", "]":"[", "}":"{"}
        
        for char in s:
            if char not in parenthesesPairs:
                parenthesesStack.append(char)
            elif not parenthesesStack or parenthesesStack[-1] != parenthesesPairs[char]:
                return False
            else:
                parenthesesStack.pop()
                
        return len(parenthesesStack) == 0

File: leetcode/0020-valid-parentheses/solution.cpp
class Solution {
    public:
        bool isValid(string s) {
            std::stack<char> stack;
            for (char ch : s){
                if (ch == '(' || ch == '[' || ch == '{') {
                    stack.push(ch);
                } else {
                    if (stack.empty()) return false;
                    
                    char top = stack.top();
                    if ((top == '(' && ch == ')') ||
                        (top == '[' && ch == ']') ||
                        (top == '{' && ch == '}')) {
                        stack.pop();
                    } else {
                        return false;
                    }
                }
            }
            return stack.empty();
        }
};